### PR TITLE
Feature/8842/add caching

### PR DESCRIPTION
**Pull request by mervyn2017**

### CODE REVIEW

File: src/main/java/com/takipi/oss/storage/resources/fs/JsonMultiFetchStorageResource.java [hunk 1]
```diff
@@ -22,27 +22,26 @@
 @Consumes(MediaType.APPLICATION_JSON)
 @Produces(MediaType.APPLICATION_JSON)
 public class JsonMultiFetchStorageResource {
-    
+
```
Comment (amithur) [inline src/main/java/com/takipi/oss/storage/resources/fs/JsonMultiFetchStorageResource.java:25]: No actual change in the whole file the whole content was reformatted (empty line tabbing was removed)...

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java [hunk 2]
```diff
@@ -16,31 +16,75 @@
...
+            if (value == null) {
+                recordsToFetch.add(recordWithData);
+            }
+            else {
+                logger.debug("Multi fetcher found key {} in cache", record.getKey());
+            }
+        }
         
-        List<RecordWithData> recordsWithData = new ArrayList<>(count);
+        final List<Runnable> tasks = new ArrayList<>(recordsToFetch.size());
```
Comment (amithur) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:72]: wouldn't it be better to put this code (until the return statement) under if (!recordsToFetch.isEmpty())?

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        memoryCache = new InMemoryCache.Builder().setMaxSize(maxSize).build();
+        cache = new Cache(memoryCache);
+    }
+
+    public String get(String key) {
+        
+        byte[] bytes = cache.get(key);
+        
+        if (bytes != null) {
+
```
Comment (amithur) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:25]: redundant empty line?

Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:25]: I added the empty line there for readability.

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java (hunk 2 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:72]: Yes, I changed it.

**Comment by amithur** [comment]
OK, changes were done on sub-branch feature/8857/speed-up-of-silver-number-downloads

**Review by amithur — APPROVED** [review]

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        
+        byte[] bytes = cache.get(key);
+        
+        if (bytes != null) {
+            
+            try {
+                return new String(bytes, "UTF-8");
+            }
+            catch (Exception e) {
+                logger.error(e.getMessage());
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:30]: Print informative message and put e in the end, something like this `logger.info("some string", e);`

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import com.takipi.oss.storage.caching.Cache;
+import com.takipi.oss.storage.caching.InMemoryCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class S3CacheImpl implements S3Cache {
+
+    private static final Logger logger = LoggerFactory.getLogger(S3CacheImpl.class);
+
+    private final InMemoryCache memoryCache;
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:12]: No need to keep the memory cache as a class member

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java [hunk 6]
```diff
@@ -16,31 +16,75 @@
...
     
     private final TaskExecutor taskExecutor;
+    private final static S3Cache cache;
+    
+    static {
+        if (maxCacheSize > 0) {
+            cache = new S3CacheImpl(maxCacheSize);
+        }
+        else {
+            cache = new S3Cache() {
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:29]: Create something like S3Cache.NOP_INSTANCE

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java [hunk 7]
```diff
@@ -16,31 +16,75 @@
 public class MultiFetcher {
     
     private static final Logger logger = LoggerFactory.getLogger(MultiFetcher.class);
+    private static long maxCacheSize = 33554432;
     
     private final TaskExecutor taskExecutor;
+    private final static S3Cache cache;
+    
+    static {
+        if (maxCacheSize > 0) {
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:25]: What does this mean? Is it typo?

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3Cache.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.takipi.oss.storage.resources.fs.multifetcher;
+
+interface S3Cache {
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3Cache.java:3]: As a role I think we shouldn't create an interface if there is only one concrete class. If one day we'll want another implementation we can always extract an interface.

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java [hunk 9]
```diff
@@ -16,31 +16,75 @@
...
+            };
+        }
+    }
     
     public MultiFetcher(TaskExecutor taskExecutor) {
         this.taskExecutor = taskExecutor;
     }
     
+    public static void setMaxCacheSize(long maxCacheSize) {
+        MultiFetcher.maxCacheSize = maxCacheSize;
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:47]: When this function called the code in the `static` section is already run.

File: src/main/java/com/takipi/oss/storage/TakipiStorageMain.java [hunk 10]
```diff
@@ -61,6 +62,7 @@ public void run(TakipiStorageConfiguration configuration, Environment environmen
 
         Filesystem filesystem = configureFilesystem(configuration, environment);
         TakipiStorageConfiguration.Multifetch multifetchConfig = configuration.getMultifetch();
+        MultiFetcher.setMaxCacheSize(multifetchConfig.getMaxCacheSize());
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/TakipiStorageMain.java:65]: Pass the cache size through the constructor of `JsonMultiFetchStorageResource`. Just the way we do for the concurrency level

File: src/main/java/com/takipi/oss/storage/caching/StringUtil.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.takipi.oss.storage.caching;
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/caching/StringUtil.java:1]: Move from caching package to some util package

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    private final Cache cache;
+
+    public S3CacheImpl(long maxSize) {
+        memoryCache = new InMemoryCache.Builder().setMaxSize(maxSize).build();
+        cache = new Cache(memoryCache);
```
Comment (davidlevanon) [inline src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:17]: Allow to add CacheLogger depending on the configuration

**Review by davidlevanon — CHANGES_REQUESTED** [review]

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3Cache.java (hunk 8 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3Cache.java:3]: There are two implementations of this interface. One is the functioning cache and the other is the dummy cache that does nothing.

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java (hunk 5 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:12]: What do you mean?

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java (hunk 12 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:17]: Currently we just use an in memory cache. We can always change it in future.

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java (hunk 6 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:29]: ok, I created a S3DummyCache class.

File: src/main/java/com/takipi/oss/storage/caching/StringUtil.java (hunk 11 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/caching/StringUtil.java:1]: done

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java (hunk 7 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:25]: If the maxCacheSize was set to zero, then use the dummy cache that does nothing.

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java (hunk 4 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/S3CacheImpl.java:30]: done

File: src/main/java/com/takipi/oss/storage/TakipiStorageMain.java (hunk 10 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/TakipiStorageMain.java:65]: done

File: src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java (hunk 9 above)
Comment (mervyn2017) [reply src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/MultiFetcher.java:47]: fixed

**Pull request closed without merging**